zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+    realm: ForeignKey[RemoteRealm]
+    token_kind: "apns" | "fcm"
+    token: str  # Provided by client.
+    push_account_id: RandomID[32] # Generated by client; need not be unique.
+    expired_time: datetime | None
+  ```
+
+- We probably also need a `zerver.models.push_notifications.PushDevice`
+  for Zulip Cloud use with no third-party bouncer? Or we could make Cloud use a bouncer.
+
+- Additionally, the push bouncer has an asymmetric key `bouncer_key`,
+  of which the public part is baked into clients.
+
+## Operation
+
+### To register a new push device for a Zulip account
+
+1. The client checks the [`/api/v1/server_settings`](https://zulip.com/api/get-server-settings)
+   to see whether the server advertises the `encrypted_push_notifications`
+   capability (or has a new enough feature level, details TBD).
+
+   If it does not, it uses the legacy push notifications registration flow.
+   (Note: We’ll likely want settings to require E2E encryption; that can come
+   after the feature is complete).
+
+1. The client generates a new RandomID `push_account_id`. It builds a JSON
+   object `push_registration` of type `PushRegistration`, and encrypts it to
+   `bouncer_key` as `encrypted_push_registration`.
+
+1. The client makes an API call to the server (authenticated with the user’s
+   API key) with arguments `push_account_id` and `encrypted_push_registration`.
+
+   The server generates a symmetric key `push_key` and creates a `PushDevice`
+   instance. The server replies to the client with `push_key`, and asynchronously
+   initiates the next step.
+
+   - Next step asynchronous: It ensures that the bouncer service 500ing will just
+     cause queueing of those requests for the local server, the mobile client
+     doesn’t need to deal with retries here. Handling of 500 is discussed below.
+
+   - TBD: Optionally, the client could include some randomness to be mixed into
+     `push_key`. We can decide if this is needed when reviewing the exact crypto
+     primitives we end up using.
+
+1. The server makes an API call to the push bouncer (authenticated with the
+   server’s API key) with arguments `user_id`, `user_uuid`, `realm_uuid`,
+   `push_account_id`, and `encrypted_push_registration`.
+
+   The push bouncer:
+   1. Decrypts `push_registration`.
+
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+      A coded `JsonableError` should be used if the timestamp is stale.
+
+   1. Checks that it doesn’t already have a `RemotePushDevice` for this `token`
+      and `push_account_id`. This check helps to avoid creating a duplicate
+      record if the client retries when a previous attempt actually succeeded.
+
+   1. Creates a `RemotePushDevice` instance, generating a `device_id`.
+
+   **On success**, the bouncer replies to the server with `device_id`.
+   The server stores `device_id` in the `PushDevice`.
+    
+   **On error**, the server keep retrying for a long time. Like maybe
```

Comment:
Not on a liveness failure. (Those don't have a hope of fixing themselves on retry.)

In that case the server should probably then just forget about that `push_account_id`​, so that the client retries from scratch.

Reply:
See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2046582369